src/storage/src/source/mysql/replication/events.rs

```diff
@@ -281,18 +291,30 @@ pub(super) async fn handle_rows_event(
             } else {
                 retractions += 1;
             }
-            container.push((data, new_gtid.clone(), diff));
+            event_buffer.push((data, new_gtid.clone(), diff));
         }
     }
 
-    // Flush this data
-    let gtid_cap = ctx.data_cap_set.delayed(new_gtid);
-    trace!(%id, "timely-{worker_id} sending container for {new_gtid:?} \
-                 with {} updates ({} additions, {} retractions) and {} \
-                 rewinds", container.len(), additions, retractions, rewind_count);
-    ctx.data_output
-        .give_container(&gtid_cap, &mut container)
-        .await;
+    // We want to emit data in individual pieces to allow timely to break large chunks of data into
+    // containers. Naively emitting data in the loop above interleaves rewinds and data emissions,
+    // which defeats a timely optimization that caches push buffers if the `.give()` time has not
```

Comment:
I think we should do it the other way around: give the normal data with `handle.give()` and buffer the rewind data which is rarer. Once all rewinds are done we can deallocate the buffer completely

Reply:
ah true, that'll use less memory!